pacman/utilities/algorithm_utilities/routing_info_allocator_utilities.py

```diff
@@ -26,39 +32,542 @@ def get_edge_groups(machine_graph, traffic_type):
+     same key
+    :param fixed_mask_groups: the group of sets of fixed masks partitions, \
+    where each set is a set of partitions which share the same key
+    :param fixed_field_groups:  the group of sets of fixed field partitions, \
+    where each set is a set of partitions which share the same key
+    :param flexi_field_groups: the group of sets of flexi field partitions, \
+    where each set is a set of partitions which share the same key.
+    :param continuous_groups: the group of sets of continuous key partitions, \
+    where each set is a set of partitions which share the same key.
+    :param none_continuous_groups: the group of sets of none continuous key \
+    partitions, where each set is a set of partitions which share the same key.
+    :param partition: The partition to place in the correct group.
+    :param partition_mappings: the mapping between partition and (set of \
+    partitions which share the same key, constraint group)
+    :rtype: None
+    """
+
+    # get types of constraints from this partition
+    (is_continuous, is_fixed_mask, is_fixed_key, is_flexi_field,
+     is_fixed_field, is_shared_key) = _check_types_of_constraints(
+        partition.constraints)
+
+    # this partition already in a share key somewhere, merge in
+    if partition in partition_mappings:
+
+        # locate previous referral
+        tracked_set, group_currently_resides_in = partition_mappings[partition]
+
+        # locate real group
+        real_group = _find_group(
+            fixed_key_groups=fixed_key_groups,
+            fixed_mask_groups=fixed_mask_groups,
+            fixed_field_groups=fixed_field_groups,
+            flexi_field_groups=flexi_field_groups,
+            continuous_groups=continuous_groups,
+            none_continuous_groups=none_continuous_groups,
+            is_fixed_mask=is_fixed_mask, is_fixed_key=is_fixed_key,
+            is_flexi_field=is_flexi_field, is_fixed_field=is_fixed_field,
+            shared_key_groups=shared_key_groups,
+            is_shared_key=is_shared_key, is_continuous=is_continuous)
+
+        # if not in the correct group, switch groups.
+        if group_currently_resides_in != real_group:
+
+            # create new tracked_set with this partition in front, as it
+            #  governed the move
+            new_tracked_set = OrderedSet()
+            new_tracked_set.add(partition)
+            for partition_to_move in tracked_set:
+                new_tracked_set.add(partition_to_move)
+
+            # remove mapping from wrong set, add to right set
+            group_currently_resides_in.remove(tracked_set)
+
+            # set tracked set to be new tracked set
+            tracked_set = new_tracked_set
+            real_group.append(tracked_set)
+
+            # update partition mapping accordingly
+            for partition in tracked_set:
+                partition_mappings[partition] = (tracked_set, real_group)
```

Comment:
This line is an example of the confusion here in - real_group contains tracked_set, but both are tracked here.  Is there a reason for this?

Reply:
real group may not be the same tracked set. if your constraints are more than just share key, and you already reside in a group from previous share key links.